code/obj/blob.dm

```diff
@@ -1067,6 +1077,51 @@
 	update_icon(override_parent)
 		return
 
+/datum/contextAction/blob_tile_upgrade
+	icon = 'icons/mob/blob_ui.dmi'
+	desc = ""
+	var/associated_ability
+	var/requires_unlock = FALSE
+
+	execute(obj/blob/blob_tile, mob/user)
+		if (QDELETED(blob_tile) || QDELETED(user))
+			return
+		if(blob_tile.type != /obj/blob)
+			return
+		blob_tile.perform_context_upgrade(src.associated_ability)
+
+	checkRequirements()
+		return TRUE
+
+	cancel
+		name = "Cancel"
+		icon_state = "blob-exit"
+
+		execute()
+			return
+
+	thick_membrane
+		name = "Build Thick Membrane"
+		icon_state = "blob-wall"
+		associated_ability = /datum/blob_ability/build/wall
+
+	fire_membrane
+		name = "Build Fire-Resistant Membrane"
+		icon_state = "blob-firewall"
+		associated_ability = /datum/blob_ability/build/firewall
+
+	slime_launcher
```

Comment:
Why is this here and not other tile upgrades like ectothermids and lipid cells? I think keeping it to just the three "wall" types would make more sense

Reply:
I was just focusing on the more common frontline/combat type cells, agree maybe slime launcher isn't needed tho